### PR TITLE
feat: allow hover providers to communicate loading state

### PR DESCRIPTION
**Pull request by felixfbecker**
This addresses https://github.com/sourcegraph/sourcegraph/issues/9349, https://github.com/sourcegraph/sourcegraph/issues/9350 and https://github.com/sourcegraph/sourcegraph/issues/9346.

Codeintellify was originally not written to support providers updating the hover content later. It accepted Observables for cancellation, but they were intended to only emit once.

With extensions, Observables started to emit multiple times to update the content. The problem is that the period between invocation and the first result is used to display a loading indicator. But with multiple emissions, the situation can arise where first an empty result is emitted (e.g. because no providers were registered yet), then providers get registered and start loading, then emit actual non-empty result. Codeintellify had no way to know that providers started loading again.

This changes the types so that when an Observable is returned, it has to contain information about when the provider is loading or not loading in addition to the result to show (through `MaybeLoadingResult<T>`). It's still possible to only emit a single result by returning a Promise, in which case this information does not need to be provided because it is known (time between invocation and resolution).

This information is then used to make the loading indicator logic smarter - it now shows a loader even if the provider had emitted an empty result before, given that the LOADER_DELAY has elapsed. This logic was also duplicated with only slight modifications for definitions (in the main repo), so it is now abstracted in its own reusable Rx operator, which has Rx marble unit tests.

PR for the main repo: https://github.com/sourcegraph/sourcegraph/pull/9914

### CODE REVIEW

File: src/loading.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    result: T
+}
+
+/**
+ * Handles input of MaybeLoadingResult (which contains both results and loading states) and returns a sequence of clear
+ * instructions on when to show a loader, results or nothing.
+ *
+ * @param loaderDelay After how many milliseconds of no results a loader should be shown.
+ * @param emptyResultValue The value that represents no results. This will be emitted, and also deep-compared to with `isEqual()`. Example: `null`, `[]`
```
Comment (lguychard) [inline src/loading.ts:29]:
```suggestion
 * @param emptyResultValue The value that represents the absence of results. This will be emitted, and also deep-compared to with `isEqual()`. Example: `null`, `[]`
```

File: src/loading.test.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
+import { TestScheduler } from 'rxjs/testing'
+import { emitLoading, LOADING, MaybeLoadingResult } from './loading'
+
+const inputAlphabet: Record<'l' | 'e' | 'r', MaybeLoadingResult<number | null>> = {
```
Comment (lguychard) [inline src/loading.test.ts:4]: I love the "shared alphabets" for tests 🙂

File: src/loading.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+     * The latest result.
+     */
+    result: T
+}
+
+/**
+ * Handles input of MaybeLoadingResult (which contains both results and loading states) and returns a sequence of clear
+ * instructions on when to show a loader, results or nothing.
+ *
+ * @param loaderDelay After how many milliseconds of no results a loader should be shown.
```
Comment (lguychard) [inline src/loading.ts:28]:
```suggestion
 * @param loaderDelay The delay, in milliseconds, after which a loader should be shown if no results have been emitted.
```

File: src/loading.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    isLoading: boolean
+
+    /**
+     * The latest result.
+     */
+    result: T
+}
+
+/**
+ * Handles input of MaybeLoadingResult (which contains both results and loading states) and returns a sequence of clear
```
Comment (lguychard) [inline src/loading.ts:25]:
```suggestion
 * Maps a stream of MaybeLoadingResult (which contains both results and loading states) to a stream of clear
```

File: src/hoverifier.ts [hunk 5]
```diff
@@ -335,12 +335,16 @@ export const TOOLTIP_DISPLAY_DELAY = 100
 export const MOUSEOVER_DELAY = 50
 
 /**
+ * Function that returns a Subscribable of the hover result to be shown.
```
Comment (lguychard) [inline src/hoverifier.ts:338]:
```suggestion
 * Function that returns a Subscribable or PromiseLike of the hover result to be shown.
```

File: src/hoverifier.ts [hunk 6]
```diff
@@ -335,12 +335,16 @@ export const TOOLTIP_DISPLAY_DELAY = 100
...
+ * It may emit on the Observable to update the content and must indicate when it starts and stopped loading new content.
```
Comment (lguychard) [inline src/hoverifier.ts:339]:
```suggestion
 * If a Subscribable is returned, it may emit more than once to update the content,
 * and must indicate when it starts and stopped loading new content.
```

File: src/loading.ts [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+    /**
+     * Whether the result provider is currently getting a new result.
+     */
+    isLoading: boolean
+
+    /**
+     * The latest result.
+     */
+    result: T
+}
```
Comment (lguychard) [inline src/loading.ts:22]: If I understood correctly the implementation of `emitLoading()` below, emissions where `isLoading === false && result !== emptyResultValue` will get ignored. Should we represent this at the type level? Eg.:

```typescript
export type MaybeLoadingResult<TResult, TEmpty> = { isLoading: boolean; result: TEmpty } | { isLoading: false; result: TResult }
```

Comment (felixfbecker) [reply src/loading.ts:22]: I think that would disallow intermediate results from the provider - e.g. in the webapp, multiple providers are invoked in parallel, and whenever one of them returns there will be an `{ isLoading: true, result: TResult }` emission as long as there are still other providers loading. At first I thought that you are right though, and that reading the logic this is (wrongly) ignored. There was no unit test for intermediate values, so I added one in https://github.com/sourcegraph/codeintellify/pull/251/commits/01322d31c7bf0869cd52aa2ae8e12e35013699f4, but to my surprise it actually passes. `{ isLoading: true, result: TResult }` passes the condition in L69 because it is a logical OR: _not loading OR not empty_ (in this case it is _not empty_), so it is not ignored but shown, but it _will_ (rightfully) not show the loader if it is the latest result at the time the loader delay elapses, because L64 makes sure to filter out non-empty results (_is loading AND not empty_ => emit `loading`, otherwise not). Does this make sense?

I'm really happy you asked about this, because it uncovered a missing test case! Thank you for being so thorough!

Comment (lguychard) [reply src/loading.ts:22]: Oh! I misread the logical OR for a logical AND 😅 This makes sense, and the use case of intermediate results is most definitely valid (and in fact I was curious how this would work _without_ handling intermediate results in the other PR). Thanks for adding the missing tests!

**Comment by felixfbecker** [comment]
Thanks for the thorough review @lguychard! I'll wait with merging this until the main repo PR is approved too so this doesn't turn out to need more adjustments after merging

File: src/loading.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+ * @template TResult The type of the provider result (without `TEmpty`).
+ * @template TEmpty The type of the empty value, e.g. `null` or `[]`.
+ */
+export const emitLoading = <TResult, TEmpty>(
+    loaderDelay: number,
+    emptyResultValue: TEmpty
+): OperatorFunction<MaybeLoadingResult<TResult | TEmpty>, TResult | TEmpty | typeof LOADING | undefined> => source => {
+    const sharedSource = source.pipe(
+        // Prevent a loading indicator to be shown forever if the source completes without a result.
+        endWith<Partial<MaybeLoadingResult<TResult | TEmpty>>>({ isLoading: false }),
```
Comment (eseliger) [inline src/loading.ts:40]: nice catch

File: src/loading.test.ts [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        scheduler.run(({ cold, expectObservable }) => {
+            const source = cold('l 10ms e', inputAlphabet)
+            expectObservable(source.pipe(emitLoading(300, null))).toBe('u 10ms e', outputAlphabet)
+        })
+    })
+    it('emits a loader if the source has not emitted after the loader delay', () => {
+        const scheduler = new TestScheduler(deepStrictEqual)
+        scheduler.run(({ cold, expectObservable }) => {
+            const source = cold('400ms r', inputAlphabet)
+            expectObservable(source.pipe(emitLoading(300, null))).toBe('u 299ms l 99ms r', outputAlphabet)
```
Comment (eseliger) [inline src/loading.test.ts:42]: Out of curiosity: what is the reason these are off by 1ms and other tests by other values? 🤔

Comment (felixfbecker) [reply src/loading.test.ts:42]: Good question :) The answer is in the rxjs docs:

> NOTE: You may have to subtract 1 millisecond from the time you want to progress because the alphanumeric marbles (representing an actual emitted value) advance time 1 virtual frame themselves already, after they emit. This can be very unintuitive and frustrating, but for now it is indeed correct.

https://rxjs-dev.firebaseapp.com/guide/testing/marble-testing#time-progression-syntax

File: src/loading.ts (hunk 8 above)
Comment (felixfbecker) [reply src/loading.ts:40]: Your comment actually made me realize there are no tests for this case, so I added more tests and realized there was a slight bug with this edge case (loader delay would still block completion). Fixed in https://github.com/sourcegraph/codeintellify/pull/251/commits/856007bb2b4109cedad1ef4f4d14852e583fc200, thanks 😊

File: src/loading.test.ts (hunk 9 above)
Comment (eseliger) [reply src/loading.test.ts:42]: Ohh wow 👀
